### PR TITLE
ImGuiIntegration example

### PR DESCRIPTION
**Pull request by Squareys**
Hello everybody!

This is the WIP PR for integrating the example from [magnum-imgui](https://github.com/lecopivo/magnum-imgui) into the main magnum repositories.

Cheers,
Jonathan

**TODOs**
 - [x] Add initial code
 - [x] Adapt codestyle
 - [x] Adapt to changes is ImGuiIntegration
 - [x] Enable on CI
 - [x] Add documentation and screenshot

### CODE REVIEW

File: src/imgui/ImGuiExample.cpp [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+#include <Magnum/Platform/Sdl2Application.h>
+#include <Magnum/GL/DefaultFramebuffer.h>
+
+#include <Magnum/ImGuiIntegration/Integration.h>
+
+#include <chrono>
+#include <thread>
+
+using namespace std::chrono_literals;
```
Comment (mosra) [inline src/imgui/ImGuiExample.cpp]: This is C++14, btw. ;)

Comment (Squareys) [reply src/imgui/ImGuiExample.cpp]: Well, I'll be using `setSwapInterval()` instead anyway ;)

File: doc/imgui.dox [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+    THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+    DEALINGS IN THE SOFTWARE.
+*/
+
+namespace Magnum {
+/** @page examples-bullet ImGUI UI
```
Comment (mosra) [inline doc/imgui.dox]: `examples-imgui` ... and find&replace all other occurences below ;) 

.. well, basically rewrite the whole thing :D

oh, and mention in `doc/example-index.dox` would be great too

File: src/imgui/ImGuiExample.cpp [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                   .setSize({1280, 960})}
+{
+    #if !defined(MAGNUM_TARGET_WEBGL) && !defined(CORRADE_TARGET_ANDROID)
+    /* Have some sane speed, please */
+    setMinimalLoopPeriod(16);
+    #endif
+}
+
+void ImGuiExample::drawEvent() {
+    GL::Renderer::setClearColor(_clearColor);
```
Comment (mosra) [inline src/imgui/ImGuiExample.cpp]: Does this need to be set every time? Is it possible to do something like `if(ImGui::ColorEdit3(...)) GL::Renderer::setClearColor()`?

Comment (Squareys) [reply src/imgui/ImGuiExample.cpp]: Yup, that's a thing 👍

**Comment by Squareys** [comment]
@mosra Ready for review! :)

File: package/ci/appveyor-desktop-gles.bat [hunk 4]
```diff
@@ -102,6 +104,7 @@ cmake .. ^
     -DWITH_BOX2D_EXAMPLE=OFF ^
     -DWITH_BULLET_EXAMPLE=ON ^
     -DWITH_CUBEMAP_EXAMPLE=OFF ^
+    -DWITH_IMGUI_EXAMPLE=OFF ^
```
[the remark this answers is not in the document]
Comment (Squareys) [reply package/ci/appveyor-desktop-gles.bat]: Probably a copy paste error 😅

File: doc/imgui.dox [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+-   @m_class{m-label m-default} **Mouse** interaction with widgets and dragging
+    windows
+-   @m_class{m-label m-default} **Keyboard** for inputting text into the UI
+
+@section examples-imgui-credits Credits
+
+This example originated from community projects by
+
+-   [\@ShaddyAQN]
```
Comment (mosra) [inline doc/imgui.dox]: ...yes, but keep the URL as well ;)

**Comment by Squareys** [comment]
Updated the find modules and rebased onto master :)

File: package/ci/travis-emscripten.sh [hunk 6]
```diff
@@ -117,6 +119,7 @@ cmake .. \
     -DWITH_BOX2D_EXAMPLE=OFF \
     -DWITH_BULLET_EXAMPLE=OFF \
     -DWITH_CUBEMAP_EXAMPLE=OFF \
+    -DWITH_IMGUI_EXAMPLE=ON \
```
Comment (mosra) [inline package/ci/travis-emscripten.sh]: Can all the `-DWITH_IMGUI` and `-DWITH_IMGUI_EXAMPLE` be `$TARGET_GLES3` like in integration? (Similarly in other ES/Android/iOS/Emscripten/WinRT scripts.)

File: package/ci/travis.yml [hunk 7]
```diff
@@ -177,6 +177,9 @@ install:
 # Bullet
 - if [ "$TRAVIS_OS_NAME" == "osx" ] && [ "$TARGET" == "desktop" ]; then HOMEBREW_NO_AUTO_UPDATE=1 brew install bullet; fi
 
+# ImGui
+- cd $HOME ; wget -nc --no-check-certificate https://github.com/ocornut/imgui/archive/v1.65.tar.gz && mkdir -p imgui && cd imgui && tar --strip-components=1 -xzf ../v1.65.tar.gz && cd $TRAVIS_BUILD_DIR
```
Comment (mosra) [inline package/ci/travis.yml]: Can this be consistent (= a copy-paste) of what's done in integration? The caching, in particular.

File: src/imgui/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    cmake_policy(SET CMP0028 NEW)
+endif()
+# Enable MACOSX_RPATH by default
+if(POLICY CMP0042)
+    cmake_policy(SET CMP0042 NEW)
+endif()
+# Quoted variables should not be dereferenced
+if(POLICY CMP0054)
+    cmake_policy(SET CMP0054 NEW)
+endif()
```
Comment (mosra) [inline src/imgui/CMakeLists.txt]: `cmake_minimum_required(VERSION 3.1)` and none of the policy setup is needed anymore.

File: package/ci/travis-android-arm.sh [hunk 9]
```diff
@@ -82,7 +82,9 @@ cmake .. \
     -DCORRADE_RC_EXECUTABLE=$HOME/deps-native/bin/corrade-rc \
     -DCMAKE_INSTALL_PREFIX=$TRAVIS_BUILD_DIR/android-ndk-r16b/platforms/android-22/arch-arm64/usr \
     -DMAGNUM_INCLUDE_INSTALL_PREFIX=$TRAVIS_BUILD_DIR/android-ndk-r16b/sysroot/usr \
+    -DIMGUI_DIR=$HOME/imgui \
     -DWITH_BULLET=OFF \
+    -DWITH_IMGUI=ON \
```
[the remark this answers is not in the document]
Comment (Squareys) [reply package/ci/travis-android-arm.sh]: Yeah, only had done that for integration and remembered examples after that 😅

**Comment by Squareys** [comment]
Alright, _now_ it should be done :D

File: src/imgui/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,73 @@
+#
+#   This file is part of Magnum.
+#
+#   Original authors — credit is appreciated but not required:
+#
+#       2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018 —
+#           Vladimír Vondruš <mosra@centrum.cz>
+#       2018 — ShaddyAQN <ShaddyAQN@gmail.com>
+#       2018 — Jonathan Hale <squareys@googlemail.com>
+#       2018 — Thomas Skrivan <skrivantomas@seznam.cz>
```
Comment (mosra) [inline src/imgui/CMakeLists.txt]: `2018 — Tomáš Skřivan <skrivantomas@seznam.cz>` 

... :laughing:

**Comment by mosra** [comment]
Merged in 37dab5b6c3447748f795681f380050ae1b5e39ee, thank you (and everyone else) for the great work! :tada: 

Since this was impossible to test for you without the integration PR being merged, I amended the commit with up-to-date FindImGui and FindMagnumIntegration as well as other minor fixes in the CI scripts. The web demo is now ready as well: https://magnum.graphics/showcase/imgui/

**Pull request closed without merging**